<a href="https://colab.research.google.com/github/Vitalyastiy/courses/blob/main/%D0%B1%D1%8B%D1%81%D1%82%D1%80%D0%BE%D0%B5%20%D0%B2%D0%B2%D0%B5%D0%B4%D0%B5%D0%BD%D0%B8%D0%B5%20%D0%B2%20pandas/10.%20%D0%9A%D0%B0%D0%BA%20%D0%BC%D0%B0%D0%BD%D0%B8%D0%BF%D1%83%D0%BB%D0%B8%D1%80%D0%BE%D0%B2%D0%B0%D1%82%D1%8C%20%D1%82%D0%B5%D0%BA%D1%81%D1%82%D0%BE%D0%B2%D1%8B%D0%BC%D0%B8%20%D0%B4%D0%B0%D0%BD%D0%BD%D1%8B%D0%BC%D0%B8%3F.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<a href="https://colab.research.google.com/github/dm-fedorov/pandas_basic/blob/master/%D0%B1%D1%8B%D1%81%D1%82%D1%80%D0%BE%D0%B5%20%D0%B2%D0%B2%D0%B5%D0%B4%D0%B5%D0%BD%D0%B8%D0%B5%20%D0%B2%20pandas/10.%20%D0%9A%D0%B0%D0%BA%20%D0%BC%D0%B0%D0%BD%D0%B8%D0%BF%D1%83%D0%BB%D0%B8%D1%80%D0%BE%D0%B2%D0%B0%D1%82%D1%8C%20%D1%82%D0%B5%D0%BA%D1%81%D1%82%D0%BE%D0%B2%D1%8B%D0%BC%D0%B8%20%D0%B4%D0%B0%D0%BD%D0%BD%D1%8B%D0%BC%D0%B8%3F.ipynb" target="_blank"><img align="left" src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open in Colab" title="Open and Execute in Google Colaboratory"></a>

<a href="https://t.me/init_python"><img src="https://dfedorov.spb.ru/pandas/logo-telegram.png" width="35" height="35" alt="telegram" align="left"></a>

In [1]:
import pandas as pd

In [2]:
url = "https://raw.githubusercontent.com/dm-fedorov/pandas_basic/master/%D0%B1%D1%8B%D1%81%D1%82%D1%80%D0%BE%D0%B5%20%D0%B2%D0%B2%D0%B5%D0%B4%D0%B5%D0%BD%D0%B8%D0%B5%20%D0%B2%20pandas/data/titanic.csv"

In [3]:
titanic = pd.read_csv(url)
titanic.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

Сделаем все имена символов строчными:

In [4]:
titanic["Name"].str.lower()

0                                braund, mr. owen harris
1      cumings, mrs. john bradley (florence briggs th...
2                                 heikkinen, miss. laina
3           futrelle, mrs. jacques heath (lily may peel)
4                               allen, mr. william henry
                             ...                        
886                                montvila, rev. juozas
887                         graham, miss. margaret edith
888             johnston, miss. catherine helen "carrie"
889                                behr, mr. karl howell
890                                  dooley, mr. patrick
Name: Name, Length: 891, dtype: object

Чтобы перевести каждую строку в столбце `Name` в нижний регистр, необходимо выбрать столбец `Name`, добавить метод `str` и применить метод [`lower`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.str.lower.html). Таким образом, каждая строка преобразуется поэлементно.

Подобно объектам `datetime`, имеющим средство доступа `dt`, при использовании `str` доступно несколько специальных строковых методов. Эти методы имеют совпадающие имена с эквивалентными встроенными строковыми методами для отдельных элементов, но применяются поэлементно для каждого из значений столбцов.

Создадим новый столбец `Surname`, содержащий фамилию пассажиров, извлекая часть перед запятой:

In [5]:
titanic["Name"].str.split(",")

0                             [Braund,  Mr. Owen Harris]
1      [Cumings,  Mrs. John Bradley (Florence Briggs ...
2                              [Heikkinen,  Miss. Laina]
3        [Futrelle,  Mrs. Jacques Heath (Lily May Peel)]
4                            [Allen,  Mr. William Henry]
                             ...                        
886                             [Montvila,  Rev. Juozas]
887                      [Graham,  Miss. Margaret Edith]
888          [Johnston,  Miss. Catherine Helen "Carrie"]
889                             [Behr,  Mr. Karl Howell]
890                               [Dooley,  Mr. Patrick]
Name: Name, Length: 891, dtype: object

Используя метод [`Series.str.split()`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.str.split.html#pandas.Series.str.split), каждое из значений возвращается в виде списка из 2 элементов. Первый элемент - это часть перед запятой, а второй элемент - часть после запятой.

In [6]:
titanic["Surname"] = titanic["Name"].str.split(",").str.get(0)

In [8]:
titanic

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked    Surname  
0        0         A/5 21171   7.2500   NaN        S     Braund  
1        0          PC 17599  71.2833   C85        C    Cumings  
2        0  STON/O2. 3101282   7.9250   NaN        S  Heikkinen  
3        0            113803  53.1000  C123        S   Futrelle  
4        0            373450   8.0500   NaN        S      Allen  
..     ...               ...      ...   ...      ...        ...  
886      0            211536  13.0000   NaN        S   Montvila  
887      0            112053  30.0000   B42        S     Graham  
888      2        W./C. 6607  23.4500   NaN        S   Johnston  
889      0            111369  30.0000  C148        C       Behr  
890      0            370376   7.7500   NaN        Q     Dooley  

[891 rows x 13 columns]

Поскольку нас интересует только первая часть, представляющая фамилию (элемент `0`), мы можем снова использовать `str`  и применить метод [`Series.str.get()`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.str.get.html#pandas.Series.str.get) для извлечения соответствующей части.

Дополнительная информация об извлечении частей строк доступна в разделе [руководства пользователя по разделению и замене строк](https://pandas.pydata.org/pandas-docs/stable/user_guide/text.html#text-split).

Получим данные о графине на борту Титаника:

In [9]:
titanic["Name"].str.contains("Countess")

0      False
1      False
2      False
3      False
4      False
       ...  
886    False
887    False
888    False
889    False
890    False
Name: Name, Length: 891, dtype: bool

In [11]:
titanic[titanic["Name"].str.contains("Countess")]

PassengerId  Survived  Pclass  \
759          760         1       1   

                                                  Name     Sex   Age  SibSp  \
759  Rothes, the Countess. of (Lucy Noel Martha Dye...  female  33.0      0   

     Parch  Ticket  Fare Cabin Embarked Surname  
759      0  110152  86.5   B77        S  Rothes

История в [Википедии](https://ru.wikipedia.org/wiki/%D0%9D%D0%BE%D1%8D%D0%BB%D1%8C_%D0%9B%D0%B5%D1%81%D0%BB%D0%B8,_%D0%B3%D1%80%D0%B0%D1%84%D0%B8%D0%BD%D1%8F_%D0%A0%D0%BE%D1%82%D0%B5%D1%81).

<img src="https://m.media-amazon.com/images/M/MV5BZTMxZmY2YzYtNDNjOS00OGE0LWFjMjQtZGRkNGU3OGMyNTQzXkEyXkFqcGdeQXVyMTk4MzEzMDY@._V1_SY1000_CR0,0,676,1000_AL_.jpg" height="300px" width="200px" >

На фото Люси Ноэль Марта Лесли, графиня Ротес, одна из выживших пассажиров затонувшего лайнера «Титаник»

Строковый метод [`Series.str.contains()`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.str.contains.html#pandas.Series.str.contains) проверяет каждое из значений в столбце, содержит ли строка слово `Countess` и возвращает `True` (если `Countess` является частью имени) или `False` (`Countess` не является частью имени). Полученные данные могут быть использованы для фильтрации с использованием условного (логического) индексирования. Поскольку на Титанике была только 1 графиня, в результате мы получаем один ряд.

Методы [`Series.str.contains()`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.str.contains.html#pandas.Series.str.contains) и [`Series.str.extract()`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.str.extract.html#pandas.Series.str.extract) поддерживают механизм [`регулярных выражений`](https://docs.python.org/3/library/re.html).

Дополнительная информация об извлечении частей строк доступна в разделе [руководства пользователя по сопоставлению и извлечению строк](https://pandas.pydata.org/pandas-docs/stable/user_guide/text.html#text-extract).

Определим, у какого пассажира самое длинное имя?

In [12]:
titanic["Name"].str.len()

0      23
1      51
2      22
3      44
4      24
       ..
886    21
887    28
888    40
889    21
890    19
Name: Name, Length: 891, dtype: int64

Чтобы получить самое длинное имя, сначала мы должны узнать длину каждого из имен в столбце `Name`, используя строковые методы `pandas`. Функция [`Series.str.len()`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.str.len.html#pandas.Series.str.len) применяется к каждому имени отдельно (поэлементно).

In [13]:
titanic["Name"].str.len().idxmax()

307

Затем необходимо получить соответствующее местоположение, желательно метку индекса в таблице, для которой длина имени самая большая. Метод [`idxmax()`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.idxmax.html) не строковый, он применяется к целым числам, поэтому не используется `str`.

In [ ]:
titanic.loc[titanic["Name"].str.len().idxmax(), "Name"]

Основываясь на индексном имени `row` (`307`) и столбце (`Name`), мы можем сделать выбор, используя оператор `loc`.

В столбце `Sex` замените значения `male` на `M`, а `female` - на `F`.

In [ ]:
titanic["Sex_short"] = titanic["Sex"].replace({"male": "M", "female": "F"})

In [ ]:
titanic["Sex_short"]

В `pandas` метод [`replace()`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.replace.html#pandas.Series.replace) предоставляет удобный способ использования отображений или словарей для замены определенных значений.

Полный обзор представлен на страницах [руководства пользователя по работе с текстовыми данными](https://pandas.pydata.org/pandas-docs/stable/user_guide/text.html#text).

<a href="https://t.me/init_python"><img src="https://dfedorov.spb.ru/pandas/logo-telegram.png" width="35" height="35" alt="telegram" align="left"></a>